spec_util/meld.go

```diff
@@ -721,37 +733,59 @@ func (m *melder) meldOneOfVariant(dst *pb.OneOf, srcHash *string, srcVariant *pb
 	// need to change the hash.
 	switch srcVariant.Value.(type) {
 	case *pb.Data_Struct:
-		// If the destination has a struct variant, merge with that. Otherwise, fall
-		// through.
-		for oldDstHash, dstVariant := range dst.Options {
+		// If the destination has a struct variant, merge with that.
+		for oldDstHash, dstVariant := range dstOneOf.Options {
 			if _, dstIsStruct := dstVariant.Value.(*pb.Data_Struct); dstIsStruct {
-				return m.meldAndRehashOption(dst, oldDstHash, dstVariant, srcVariant)
+				return m.meldAndRehashOption(dstOneOf, oldDstHash, dstVariant, srcVariant)
 			}
 		}
 
 	case *pb.Data_List:
-		// If the destination has a list variant, merge with that. Otherwise, fall
-		// through.
-		for oldDstHash, dstVariant := range dst.Options {
+		// If the destination has a list variant, merge with that.
+		for oldDstHash, dstVariant := range dstOneOf.Options {
 			if _, dstIsList := dstVariant.Value.(*pb.Data_List); dstIsList {
-				return m.meldAndRehashOption(dst, oldDstHash, dstVariant, srcVariant)
+				return m.meldAndRehashOption(dstOneOf, oldDstHash, dstVariant, srcVariant)
 			}
 		}
 
 	case *pb.Data_Primitive:
-		// If the destination has a primitive variant that has a compatible type
-		// with the source variant, meld them.  Otherwise, fall through.
-		for oldDstHash, dstVariant := range dst.Options {
+		// The destination might have more than one primitive variant whose type is
+		// compatible with the source variant. If the source variant has no data
+		// formats (i.e., is just a base type), then all these variants will
+		// collapse down to a single variant when melded. Therefore, using
+		// srcVariant as an accumulator, we remove all compatible variants from the
+		// destination and meld them together with the source variant. Later on, we
+		// will add the resulting source variant to the destination.
+		needToRehashSrc := false
+		for dstHash, dstVariant := range dstOneOf.Options {
 			if _, dstIsPrim := dstVariant.Value.(*pb.Data_Primitive); dstIsPrim && haveCompatibleTypes(dstVariant.GetPrimitive(), srcVariant.GetPrimitive()) {
-				return m.meldAndRehashOption(dst, oldDstHash, dstVariant, srcVariant)
+				delete(dstOneOf.Options, dstHash)
+
+				// To avoid modifying *srcVariant, we meld srcVariant into dstVariant,
+				// and then set srcVariant = dstVariant.
+				if err := m.meldData(dstVariant, srcVariant); err != nil {
+					return err
+				}
+				srcVariant = dstVariant
+				needToRehashSrc = true
 			}
 		}
 
+		if needToRehashSrc {
+			srcHash = ir_hash.HashDataToString(srcVariant)
+		}
+
 	default:
 		return fmt.Errorf("unknown one-of variant type: %s", reflect.TypeOf(srcVariant.Value).Name())
 	}
 
+	if len(dstOneOf.Options) == 0 {
+		// No longer have a one-of. Replace dst's value with the singleton variant.
+		dst.Value = srcVariant.Value
```

Comment:
I am trying to think through all the cases that this new code might be triggered.  Is there a case where we need to re-hash dst where we would not have re-hashed previously?  (I don't think so, we should not have nested OneOf's.)

This new case seems to be exclusive to primitives, should we move it inside the case?  Or do you think this check is more broadly applicable?

Reply:
I'll think about the first question, but to answer the second, I say we just leave this here. A singleton one-of isn't really a one-of and should be collapsed.

Reply:
According to [this](https://github.com/akitasoftware/akita-libs/blob/c49ba8585c347d1927a2dd78fc9516d9807d7f19/spec_util/hash_keys.go#L9-L12), there are only three places in the IR where hashes are used: `Method.Args`, `Method.Responses`, and `OneOf.Options`. Any rehashing for the first two cases should already be handled as part of the contract on the meld procedure. And the third, as you say, shouldn't happen because we don't nest one-ofs.